src/pybind/rbd/rbd.pyx

```diff
@@ -392,6 +398,69 @@ cdef int progress_callback(uint64_t offset, uint64_t total, void* ptr) with gil:
 cdef int no_op_progress_callback(uint64_t offset, uint64_t total, void* ptr):
     return 0
 
+
+# Python 3.12+ callback wrapper for sub-interpreter safety
+cdef class ProgressCallbackWrapper:
+    """
+    Wraps progress callbacks for Python 3.12+ sub-interpreter safety.
+
+    Python 3.12+ introduced significant changes to sub-interpreter isolation
+    which can cause crashes when callbacks are invoked from C code in a
+    different interpreter context than where the callback was created.
+
+    This wrapper stores the interpreter state when created and switches
+    to the correct interpreter when the callback is invoked, ensuring
+    the callback executes in its original interpreter context.
```

Comment:
In addition to `on_progress` callbacks, RBD has a number of AIO/asynchronous methods that take `oncomplete` callbacks: `RBD.aio_open_image()`, `Image.aio_close()`, `Image.aio_read()`, etc.  At least some of them also seem to misbehave on Python 3.12+ with sub-interpreters: a trivial reproducer for `Image.aio_mirror_image_get_info()` is captured in https://tracker.ceph.com/issues/73857.  Although ceph-mgr daemon doesn't crash there, the call to `Image.aio_mirror_image_get_info()` executed from such a callback for `RBD.aio_open_image()` effectively gets ignored.

I see that you have added a relation from the `on_progress` tracker to https://tracker.ceph.com/issues/73857.  Do you have a similar kind of fix in mind?